lib/typecheck.ts

```diff
@@ -392,1195 +421,1293 @@ async function typeCheckFilter(ast : Ast.BooleanExpression,
+
+        if (ast.actions.some((a) => a.isNotify)
+            && (ast instanceof Ast.Command || ast instanceof Ast.OnInputChoice)
+            && !ast.table)
+            throw new TypeError('Cannot return a result without a GET function');
+
+        for (const prim of ast.actions)
+            this._addRequiredInputParamsAction(prim);
+        await Promise.all(
+            ast.actions.map((action) => this._typeCheckAction(action, scope)));
+    }
+
+    private async _typeCheckExpressionStatement(ast : Ast.ExpressionStatement,
+                                                scope : Scope) {
+        await this._loadAllSchemas(ast, scope);
+
+        const expressions = ast.expression.expressions;
+        for (let i = 0; i < expressions.length; i++) {
+            const expr = expressions[i];
+            this._addRequiredInputParamsExpression(expr);
+
+            scope.$has_event = i > 0;
+            await this._typeCheckExpression(expr, scope);
+            const schema = expr.schema!;
+            if (schema.require_filter)
+                throw new TypeError('Filter required');
+            if (schema.functionType === 'stream' && i > 0)
+                throw new Error(`Stream expression must be first in a chain expression`);
+            if (schema.functionType === 'action' && i < expressions.length-1)
+                throw new Error(`Action expression must be last in a chain expression`);
+        }
+        this._resolveChain(ast.expression);
     }
 
-    if (!ast.isAction) {
-        await loadAllSchemas(ast, schemas, scope, classes, useMeta);
+    private _typeCheckProgramAnnotations(ast : Ast.Program2) {
+        return Promise.all(Object.entries(ast.impl_annotations).map(async ([name, value] : [string, Ast.Value]) => {
+            if (!value.isConstant())
+                throw new Error(`Annotation #[${name}] must be a constant`);
 
-        // if isAction was `undefined` (before typechecking) overwrite it to `false`
-        ast.isAction = false;
+            if (name === 'principal')
+                await this._typecheckPrincipal(value);
+        }));
     }
 
-    addRequiredInputParamsTable(ast.value, new Set<string>());
-    await typeCheckTable(ast.value, schemas, scope, classes, useMeta);
+    async typeCheckProgram2(ast : Ast.Program2,
+                            parentScope : Scope|null = null) : Promise<void> {
+        ast.classes.forEach((ast) => {
+            this._classes[ast.name] = ast;
+        });
 
-    // remove all input parameters (which we have filled with $undefined)
-    const args = Array.from(ast.value.schema!.iterateArguments()).filter((a) => !a.is_input);
-    ast.schema = new Ast.FunctionDef(ast.location, 'query', null, ast.name, [],
-        { is_list: ast.value.schema!.is_list, is_monitorable: false }, args, {});
-    scope.addGlobal(ast.name, ast.schema);
-}
+        this._typeCheckMetadata(ast);
+        this._typeCheckProgramAnnotations(ast);
```

Comment:
`this._typeCheckProgramAnnotations` should be `async` and we should need `await` here?